content: blog: Zephyr memory layout update (#356)
espressif/developer-portal: content/blog/esp32-bootstrapping/index.md

Additional information on the flash and/or SPIRAM regions are provided as a linker symbols and are embedded inside the application image.

## Building

Before we build some example applications, let’s take a moment to discuss how the linking process utilizes SRAM on the ESP32-S3.
It’s important to note that different Espressif SoCs can have unique memory layouts, which can affect how resources are allocated. Therefore each SoC is using slightly different linker script.

### Memory utilization

Following images illustrates how the application linking is done. Highlighted are details of ROM code usage in SRAM. I-Cache and D-Cache allocations are highlighted with red color.

> Important linker symbols are in green bubbles. 

Reviewer: Edit leftovers?

Author: yep, those are leftovers. Thx!

The images below illustrate the process of application linking, focusing on how the ROM code interacts with SRAM. Key details, such as I-Cache and D-Cache allocations, are highlighted in red for clarity.
Important linker symbols, crucial to understanding the memory layout, are shown in green bubbles.
These visuals are designed to make the relationships between different components and memory allocations easier to follow.

#### ESP32-S3 use-case

> Here we are using the ESP32-S3 as reference platform to illustrate the memory utilization, which is **suitable as an reference for majority of newer SoCs**.

Reviewer: How about this?

```suggestion
Here we are using the ESP32-S3 to illustrate memory utilization, which is **suitable as a reference for most of the newer SoCs**.
```

> - Yellow area is the memory used by the 2nd stage bootloader and can be re-claimed by the application code run-time.

Reviewer: ```suggestion
- Yellow area is the memory used by the 2nd stage bootloader that can be re-claimed by the application code run-time.
```

- Orange area is the memory used by the 1st stage bootloader (or ROM-loader), that can be re-claimed after the application loading is done.
- Red area is the memory that is not available for the user and should not be used by the linker to spill the code or data.

The following picture illustrates the memory utilization for an single CPU scenario.

{{< figure
    default=true
    src="img/esp32s3-zephyr-memory-default.webp"
    alt=""
    caption="The ESP32-S3 'default' memory utilization."
    >}}

The following picture illustrates the memory utilization for a multi CPU scenario.

{{< figure
    default=true
    src="img/esp32s3-zephyr-memory-amp.webp"
    alt=""
    caption="The ESP32-S3 'AMP' memory utilization."
    >}}

***NOTE:***
*The I-Cache allocation SRAM blocks (0,1) are set by the **`PMS_INTERNAL_SRAM_ICACHE_USAGE`** bits
and the D-Cache allocation SRAM blocks (9,10) are set by the **`PMS_INTERNAL_SRAM_DCACHE_USAGE`** bits, both in the register `PMC_INTERNAL_SRAM_USAGE_1_REG`*

#### ESP32 use-case

> Here is the memory utilization for the ESP32 platform as its memory model is significantly different from other SoCs in the ESP line.

Reviewer: _ESP line_ is something new. Let's stick to either _Espressif SoCs_ or _ESP chips_?

```suggestion
Here is the memory utilization for the ESP32 platform as its memory model is significantly different from other Espressif SoCs.
```

> Following illustration is covering an single CPU scenario.

Reviewer: ```suggestion
The following picture illustrates memory utilization in a single CPU scenario.
```

{{< figure
    default=true
    src="img/esp32-zephyr-memory-default.webp"
    alt=""
    caption="The ESP32 'default' memory layout."
    >}}

> Following picture illustrates the memory utilization in the multi CPU scenario.

Reviewer: ```suggestion
The following picture illustrates memory utilization in a multi-CPU scenario.
```

uart:~$
```

### MCUboot Espressif port (EP)

To learn how to build the MCUboot Espressif Port, check out this [article](https://docs.mcuboot.com/readme-espressif.html)

### AMP enabled sample code

> AMP builds require several images to be build and flash into a target SoC. Lets use ESP32-S3 as out test platform and demonstrate the sysbuild capabilities on the sample code that is provided by the Zephyr RTOS sources.

Reviewer: ```suggestion
AMP builds require several images to be built and flashed onto a target SoC. Let's use ESP32-S3 as our test platform and demonstrate the sysbuild capabilities on the sample code provided by the Zephyr RTOS sources.
```

> The sample code uses IPM (Inter-Processor Mailbox) to demonstrate simple two way communication between the PROCPU core and APPCPU core. Images for both CPU cores must be loaded using the MCUboot. Note, there is no support to run AMP while using simple boot mechanism.

Reviewer: Not clear what sample code is meant here.

```suggestion
The sample code uses IPM (Inter-Processor Mailbox) to demonstrate a simple two-way communication between the `PRO_CPU` core and `APP_CPU` core. The images for both CPU cores must be loaded using MCUboot. Note that there is no support for running AMP if the Simple Boot mechanism is used.
```

```shell
cd zephyrproject/zephyr
west build -b esp32s3_devkitm/esp32s3/procpu samples/drivers/ipm/ipm_esp32/ -p --sysbuild
west flash && west espressif monitor

> ```

Reviewer: Please introduce briefly what these commands do.

> Three images has been created. Image for the PROCPU is called `ipm_esp32`. Image for the APPCPU is called `ipm_esp32_remote`. Finally MCUboot image (run by PROCPU) is called `mcuboot`.

Reviewer: ```suggestion
As a result, three images should be created:

 - `ipm_esp32` - the image for the `PRO_CPU`
 - `ipm_esp32_remote` - the image for the `APP_CPU`
 - `mcuboot` - the MCUboot image, which is run by `PRO_CPU`
```

Author: good one

> After flashing and connecting to a target using the serial port you should be able to observe following output on the console.

Reviewer: Is it worth providing instructions or some reminders here on how to flash and connect to the serial port? If the readers are supposed to know it already, then please ignore my question.

```suggestion
After flashing and connecting to a target using the serial port, we should be able to see the following output in the console:
```

Author: the flash instructions are part of the code block describing a build

uart:~$
```

### MCUboot Espressif port (EP)

To learn how to build the MCUboot Espressif Port, check out this [article](https://docs.mcuboot.com/readme-espressif.html)

### AMP enabled sample code

AMP builds require several images to be built and flashed onto a target SoC. Let's use ESP32-S3 as our test platform and demonstrate the sysbuild capabilities on the sample code provided by the Zephyr RTOS sources.
The Zephyr sample code (`samples/drivers/ipm/ipm_esp32`) uses IPM (Inter-Processor Mailbox) to demonstrate a simple two-way communication between the `PRO_CPU` core and `APP_CPU` core. Images for both CPU cores must be loaded using the MCUboot.
Note that there is no support for running AMP if the Simple Boot mechanism is used.

> Building and flashing a complete set of images running the IPM sample:

Reviewer: How about this:

```suggestion
Now we can build and flash a complete set of images running the IPM sample:
```

Author: thx! fixed

Author: Thanks @f-hollow, I've fixed all remaining findings.
